File: app/src/main/java/projects/csce/evence/view/ui/QRDialog.java
package projects.csce.evence.view.ui;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.provider.CalendarContract;
import android.view.LayoutInflater;

import androidx.databinding.DataBindingUtil;

import projects.csce.evence.R;
import projects.csce.evence.databinding.DialogBoxQrBinding;
import projects.csce.evence.service.model.FileManager;
import projects.csce.evence.view.ui.model.ViewCalendarData;
import projects.csce.evence.view.ui.model.ViewEvent;

import static android.content.Intent.ACTION_INSERT;

public class QRDialog
{
    private static final String TAG = "QRDialog";
    private Context context;
    private Dialog dialog;
    private ViewCalendarData ical;
    private ViewEvent currentEvent;
    DialogBoxQrBinding binding;
    private FileManager fileManager;


    public QRDialog(Context context, ViewCalendarData ical, FileManager fileManager) {
        this.context = context;
        this.ical = ical;
        this.currentEvent = ical.getEvents().get(0);
        binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.dialog_box_qr, null, false);
        this.fileManager = fileManager;
        binding.setView(this);

        binding.qrDialogEventTitleTextview.setText(currentEvent.getTitle());
        binding.qrDialogEventStartDateTextview.setText(currentEvent.getStartDate());
        binding.qrDialogEventStartTimeTextview.setText(currentEvent.getStartTime());
        binding.qrDialogEventLocationTextview.setText(currentEvent.getLocation());
        binding.qrDialogQrImageview.setImageBitmap(currentEvent.getImage());
        binding.editBtn.setOnClickListener(view -> {
            Intent intent = new Intent(context, GenerateQR.class);
            intent.putExtra("FILE_PATH", fileManager.getFilePath(ical.getFileName()));
            context.startActivity(intent);
        });

        dialog = new Dialog(context);
        dialog.setContentView(binding.getRoot());
        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.setCancelable(true);
        dialog.setCanceledOnTouchOutside(true);
        dialog.show();
    }

    public void save( ) {
        // Bug the user about storing it
        Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("application/*");
        intent.putExtra(Intent.EXTRA_TITLE, currentEvent.getTitle() + ".ics");

        if (context instanceof GenerateQR) {
            ((GenerateQR) context).startActivityForResult(intent, 1);
        } else {
            ((MainActivity) context).startActivityForResult(intent, 1);
        }
    }

    public void shareQR() {
        Intent shareIntent = new Intent();
        shareIntent.setAction(Intent.ACTION_SEND);
        shareIntent.putExtra(Intent.EXTRA_STREAM, fileManager.getFileUri("image_" + ical.getFileName()));
        shareIntent.setType("image/*");
        context.startActivity(Intent.createChooser(shareIntent, "Share images to.."));
    }

    public void importToCalendar() {
        Intent toCalendar = new Intent(ACTION_INSERT);
        toCalendar.setData(CalendarContract.Events.CONTENT_URI);
        toCalendar.putExtra(CalendarContract.Events.TITLE, currentEvent.getTitle());
        toCalendar.putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, currentEvent.getStartInstantEpoch());
        toCalendar.putExtra(CalendarContract.EXTRA_EVENT_END_TIME, currentEvent.getEndEpochMilli());
        toCalendar.putExtra(CalendarContract.Events.EVENT_LOCATION, currentEvent.getLocation());
        toCalendar.putExtra(CalendarContract.Events.DESCRIPTION, currentEvent.getDescription());
        context.startActivity(toCalendar);
    }
}

File: QRDialogView/src/main/java/teamevence/evenceapp/qrdialogview/QRDialog.kt
package teamevence.evenceapp.qrdialogview

import android.app.ActionBar
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.provider.CalendarContract
import android.view.LayoutInflater
import android.view.WindowManager
import coil.ImageLoader
import coil.load
import com.jakewharton.rxbinding4.view.clicks
import teamevence.evenceapp.core.*
import teamevence.evenceapp.core.db.EventOps
import teamevence.evenceapp.core.model.ViewEvent
import teamevence.evenceapp.qrdialogview.databinding.DialogBoxQrBinding
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import timber.log.Timber
import java.util.concurrent.TimeUnit

class QRDialog(
        private val context: Context,
        private val event: ViewEvent,
        private val imageLoader: ImageLoader,
        private val activityStarter: ActivityStarter,
        private val eventOps: EventOps
        )
{
    private val disposables = CompositeDisposable()

    private val binding: DialogBoxQrBinding = DialogBoxQrBinding.inflate(LayoutInflater.from(context)).apply {
        qrDialogEventTitleTextview.text = event.title
        qrDialogEventStartDateTextview.text = event.startDatePretty()
        qrDialogEventStartTimeTextview.text = event.startDateTime.toAMPM()
        qrDialogEventLocationTextview.text = event.location
        qrDialogQrImageview.load(event.imageFileUri, imageLoader)
    }

    private val dialog: Dialog = Dialog(context).apply {
        setContentView(binding.root)
        window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        setCancelable(true)
        setCanceledOnTouchOutside(true)
        window?.setLayout(WindowManager.LayoutParams.MATCH_PARENT, ActionBar.LayoutParams.MATCH_PARENT)
        Timber.i("=qrdialog context" +  context.toString())
        show()
    }

    init {
        setupClicks()
    }

    private fun setupClicks() {
        binding.closeDialogBtn.setOnClickListener { closeDialog() }
        binding.shareQrBtn.setOnClickListener {  shareQR() }
        binding.importToCalendarBtn.setOnClickListener {  importToCalendar() }
        binding.saveBtn.setOnClickListener {  save() }
        binding.editBtn.setOnClickListener {
            activityStarter.startEditQr(context, event)
            closeDialog()
        }

        disposables += binding.deleteBtn.clicks()
                .map { event.id }
                .debounce(300, TimeUnit.MILLISECONDS)
                .flatMap { id ->
                    Observable.just(Unit)
                            .map { eventOps.deleteById(id) }
                }
                .observeOn( AndroidSchedulers.mainThread() )
                .subscribe { closeDialog() }

    }

    private fun save() {
        // Bug the user about storing it
        val intent = Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = "text/calendar"
            putExtra(Intent.EXTRA_TITLE, event.title + ".ics")
        }


        if (context is BaseActivity) {
            context.startActivityForResult(intent, RequestCodes.REQUEST_SAF)
        }
    }

    private fun shareQR() {
        val shareIntent = Intent().apply {
            action = Intent.ACTION_SEND
            putExtra(Intent.EXTRA_STREAM, event.imageFileContentUri)
            type = "image/png"
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
        context.startActivity(Intent.createChooser(shareIntent, "Share images to.."))
    }

    private fun importToCalendar() {
        val toCalendar = Intent(Intent.ACTION_INSERT)
        toCalendar.data = CalendarContract.Events.CONTENT_URI
        toCalendar.putExtra(CalendarContract.Events.TITLE, event.title)
        toCalendar.putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, event.startEpochMilli())
        toCalendar.putExtra(CalendarContract.EXTRA_EVENT_END_TIME, event.endEpochMilli())
        toCalendar.putExtra(CalendarContract.Events.EVENT_LOCATION, event.location)
        toCalendar.putExtra(CalendarContract.Events.DESCRIPTION, event.description)
        context.startActivity(toCalendar)
    }

    private fun closeDialog()
    {
        // Close the Dialog and the activity
        if (context is DialogClosable)
        {
            dialog.dismiss()
            (context as DialogClosable).close()
        }
        // Just close the Dialog
        else
        {
            dialog.dismiss()
        }

        disposables.dispose()
    }

    companion object {
        private const val TAG = "QRDialog"
    }

}